### PR TITLE
Update free-programming-books-subjects.md

### PR DESCRIPTION
**Pull request by TimTeitelbaum**
I am the author of a new language-agnostic introduction-to-programming textbook, discovered this site, and would like to add my book. The use case I am thinking of is this:

A kid wants to learn to program, doesn't know which in language, and comes to this:

     BY PROGRAMMING LANGUAGE
     Originally, this list included a section called "Language Agnostic" for books about
      programming subjects not restricted to a specific programming language. That 
      section got so big, we decided to split it into its own file, the BY SUBJECT file.

The kid in intrigued by the idea of learning programming in a language-agnostic manner, goes to BY SUBJECT, and but finds there a list of specialized and advanced-sounding topics. The kid just wants to learn to program, and not finding a suitable category in which to look, is encouraged thereby to return to BY PROGRAMMING LANGUAGE, pick a specific language, and go down that path.

Little did the kid suspect that there are language-agnostic introductory-programming texts in Theoretical Computer Science. It took Eric Hellman to point out to me that this is where such great language-agnostic introductory-programming texts as SICP/Ableson&Sussman are catalogued (Yes, SICP uses Scheme, but it isn't really about "How to Program in Scheme"). But the kid will never find SICP, because "Theoretic Computer Science" is an intimidating misnomer. (And furthermore, the category does include highly-theoretical books.) The kid has lost an opportunity, and has been led to believe that programming is all about learning one of the standard programming languages.

I am proposing to:

1. Rename the category "Language-Agnostic Introductory Programming / Theoretical Computer Science".
2. Move into alphabetical order in the list of subjects.
3. Insert my own book there.

My preference (frankly) would be to separate introductory programming textbooks from highly theoretical books, but (certainly, as a newcomer to the site) I don't propose such a step.

## What does this PR do?
Add resource(s) | Remove resource(s) | Add info | Improve repo

## Checklist:
- [x] Read our [contributing guidelines](https://github.com/EbookFoundation/free-programming-books/blob/main/docs/CONTRIBUTING.md).
- [x] [Search](https://ebookfoundation.github.io/free-programming-books-search/) for duplicates.
- [x] Include author(s) and platform where appropriate.
- [x] Put lists in alphabetical order, correct spacing.
- [x] Add needed indications (PDF, access notes, under construction).
- [x] Used an informative name for this pull request.

## Follow-up

- Check the status of GitHub Actions and resolve any reported warnings!

### CODE REVIEW

**Comment by eshellman** [comment]
1. I agree that the section heading is no longer a good fit. But your suggestion is long and esoteric- not one that would occur to the target audience. How about just "Programming"
2. Please separate the PR into 2PRs - one to add your book, (an easy add) and one to reorg the section names - might take more discussion.

**Comment by TimTeitelbaum** [comment]
> 2. Please separate the PR into 2PRs - one to add your book, (an easy add) and one to reorg the section names - might take more discussion.
     Will do.
> 1. I agree that the section heading is no longer a good fit. But your suggestion is long and esoteric- not one that would occur to the target audience. How about just "Programming"

- "Language-Agnostic" is certainly esoteric for my hypothetical kid, but I was trying to keep an association with the use of the term in the forwarding reference of the preamble of "By Subject". "Programming" works as a catch-all for the books about programming, but then leaves the theoretical books out in the cold.

- I had been trying to avoid major surgery on the section "Theoretical Computer Science", but decided to take a whack at a better categorization. In turned out to not be as hard as I thought it might be, and in the end I don't think what I suggest below is all that debatable.

If there is agreement, I would leave the Theoretical Computer Science category name alone, but create a new category "Programming" into which I would move all books labeled "PROGRAMMING" below.

The current books in "Theoretical Computer Science" are:

    [Building Blocks for Theoretical Computer Science](http://mfleck.cs.illinois.edu/building-blocks/index.html) - Margaret M. Fleck
        THEORY
    [Category Theory for Computing Science](http://www.tac.mta.ca/tac/reprints/articles/22/tr22.pdf) (PDF)
        THEORY
    [Category Theory for Programmers](https://github.com/hmemcpy/milewski-ctfp-pdf) - Bartosz Milewski (PDF)
        THEORY
    [Delftse Foundations of Computation](https://textbooks.open.tudelft.nl/textbooks/catalog/book/13) - Stefan Hugtenburgand, Neil Yorke-Smith @ TU Delft Open (PDF)
        THEORY
    [Design of a Programmer](https://www.smashwords.com/books/view/639609) - Prakash Hegade (PDF)
        PROGRAMMING
    [Homotopy Type Theory: Univalent Foundations of Mathematics](http://homotopytypetheory.org/book/) (PDF)
        THEORY
    [Introduction to Computer Science](http://www.cse.iitd.ernet.in/~suban/CSL102/) - Prof. Subhashis Banerjee, IIT Delhi
        PROGRAMMING
    [Introduction to Computing](http://www.computingbook.org/)
        PROGRAMMING
    [Introduction to Theory of Computation](http://cglab.ca/~michiel/TheoryOfComputation/) - Anil Maheshwari, Michiel Smid (PDF)
        THEORY
    [Models of Computation](http://cs.brown.edu/people/jes/book/) - John E. Savage
        THEORY
    [Practical Foundations for Programming Languages, Preview](https://www.cs.cmu.edu/~rwh/pfpl/2nded.pdf) - Robert Harper (PDF)
        THEORY
    [Principles of Programming Languages](https://web.archive.org/web/20150418034451/http://www.cs.jhu.edu/~scott/pl/book/dist/) - Scott F. Smith
        THEORY
    [Programming and Programming Languages](https://papl.cs.brown.edu/2019/) - Shriram Krishnamurthi
        PROGRAMMING
    [Programming in Martin-Löf's Type Theory](http://www.cse.chalmers.se/research/group/logic/book/) - Bengt Nordstroem
        THEORY
    [Programming Languages: Application and Interpretation (2nd Edition)](http://cs.brown.edu/~sk/Publications/Books/ProgLangs/) - Shriram Krishnamurthi
        PROGRAMMING
    [Programming Languages: Theory and Practice](http://people.cs.uchicago.edu/~blume/classes/aut2008/proglang/text/offline.pdf) - Robert Harper (PDF)
        THEORY
    [Semantics with Applications: A Formal Introduction](http://www.cs.ru.nl/~herman/onderwijs/semantics2019/wiley.pdf) - Hanne Riis Nielson, Flemming Nielson (PDF)
        THEORY
    [Structure and Interpretation of Computer Programs](https://web.mit.edu/6.001/6.037/sicp.pdf) - Harold Abelson, Gerald Jay Sussman, Julie Sussman (PDF)
        PROGRAMMING
    [Structure and Interpretation of Computer Programs](https://sarabander.github.io/sicp/html/index.xhtml) - Harold Abelson, Gerald Jay Sussman, Julie Sussman (HTML)
        PROGRAMMING
    [The Black Art of Programming](http://self.gutenberg.org/wplbn0002828847-the-black-art-of-programming-by-mcilroy-mark.aspx?) - Mark McIlroy
        PROGRAMMING
    [The Craft of Programming](https://kilthub.cmu.edu/articles/The_Craft_of_Programming/6610514) - John C. Reynolds
        PROGRAMMING
    [The Nature of Code](https://natureofcode.com/book) - Daniel Shiffman (HTML)
        PROGRAMMING
    [Theory of Computation, Making Connections](https://joshua.smcvt.edu/computation) - Jim Hefferon (PDF)
        THEORY
    [Think Complexity](https://greenteapress.com/wp/think-complexity-2e/) - - Allen B. Downey (2nd Edition) (PDF, HTML)
        PROGRAMMING

- To better connect this up with the preamble of "By Programming Languages", I would propose to add the sentence in brackets (but without the brackets).

    Originally, this list included a section called "Language Agnostic" for books about programming subjects not restricted to a specific programming language. That section got so big, we decided to split it into its own file, the [BY SUBJECT file](https://github.com/EbookFoundation/free-programming-books/blob/main/books/free-programming-books-subjects.md).  [Books on general-purpose programming that don't focus on a specific language are in Programming. Books on specialized topics appear in their own sub-sections.]

**Comment by TimTeitelbaum** [comment]
**HELP**. I'm sorry, but I am sort of GitHub lost:
(1) As per Eric's suggestion, I went back to my fork, edited the By Subjects and By Languages files, and committed them. However, those commits failed Lint checks.
(2) I then reedited those files, and re-committed them, which seems to have succeeded.
(3) However, I then looked at the pending PR, and although I see a record there of my latest (successful) commits, I also see a record of the earlier (failed) commits. 
==> What I don't understand is whether there is something I need to do now (e.g., to free something that is blocking the PR), or whether I just need to wait for the "pig to pass through the python".

**Comment by eshellman** [comment]
just look at the "details" from the failing action. :
` 741:1-752:118  warning  Incorrect number of blank lines between last section and next heading  blank-lines-1-0-2  remark-lint`

**Comment by TimTeitelbaum** [comment]
> just look at the "details" from the failing action. :
>  741:1-752:118  warning  Incorrect number of blank lines between last section and 
>  next heading  blank-lines-1-0-2  remark-lint
I did so, understood the mistake, and created corrected versions of the files. I then did commits that, as far as I can tell, passed lint checks (if in fact, they were re-initiated).
v
These are shown in PR message thread as having occurred about 16 hours ago, as the very next thing after your "Excellent idea. Make it so" remark.

What I don't know is what happens when a  subsequent commit to the same file as failed a commit is then recommitted without error. Perhaps I did this in an incorrect manner.  My impression is that the earlier failed commit is somehow blocking the PR, but I don't know what to do about it. This may be an incorrect diagnosis.

I see this:
> Add more commits by pushing to the [patch-1](https://github.com/TimTeitelbaum/free-programming-books/tree/patch-1) branch on [TimTeitelbaum/free-programming-books](https://github.com/TimTeitelbaum/free-programming-books).
but don't know what it means. Is "pushing" something I should have understood and done?

I think I have correct versions of the files committed in my fork, but don't know how to now push them through in this PR. One possibility that has occurred to me to just kill this PR, and do another, but I'm not sure how to do that. Also, I would think there must be a way to push this one through, and retain the association with the message threat that is here.

File: books/free-programming-books-subjects.md [hunk 1]
```diff
@@ -735,6 +736,21 @@ Kerridge (PDF) (email address *requested*, not required)
...
+* [Programming and Programming Languages](https://papl.cs.brown.edu/2019/) - Shriram Krishnamurthi
+* [Programming Languages: Application and Interpretation (2nd Edition)](http://cs.brown.edu/~sk/Publications/Books/ProgLangs/) - Shriram Krishnamurthi
+* [Structure and Interpretation of Computer Programs](https://web.mit.edu/6.001/6.037/sicp.pdf) - Harold Abelson, Gerald Jay Sussman, Julie Sussman (PDF)
+* [Structure and Interpretation of Computer Programs](https://sarabander.github.io/sicp/html/index.xhtml) - Harold Abelson, Gerald Jay Sussman, Julie Sussman (HTML)
+* [The Black Art of Programming](http://self.gutenberg.org/wplbn0002828847-the-black-art-of-programming-by-mcilroy-mark.aspx?) - Mark McIlroy
+* [The Craft of Programming](https://kilthub.cmu.edu/articles/The_Craft_of_Programming/6610514) - John C. Reynolds
+* [The Nature of Code](https://natureofcode.com/book) - Daniel Shiffman (HTML)
+* [Think Complexity](https://greenteapress.com/wp/think-complexity-2e/) - - Allen B. Downey (2nd Edition) (PDF, HTML)
+
```
Comment (eshellman) [inline books/free-programming-books-subjects.md:753]: add a space here

**Pull request merged**